### PR TITLE
Return address and images on GET /companies/:id endpoint.

### PR DESCRIPTION
**Pull request by laurandarius**

### CODE REVIEW

**Comment by laurandarius** [comment]
[Return address and images on GET /companies/:id endpoint.](https://trello.com/c/GUg2CPBA/20-return-address-and-images-on-get-companies-id-endpoint)

File: app/services/company.js [hunk 1]
```diff
@@ -3,30 +3,93 @@ const path = require('path');
...
+const fieldOfActivityRepository = require('../repositories/fieldOfActivity');
 const fileService = require('./file');
 const error = require('../services/error');
 const CrudService = require('./crudService');
 const companyFilesFolder = require('../../config.js').companyFilesFolder;
+const ORIGIN = require('../../config.js').ORIGIN;
 
 class CompanyService extends CrudService {
+  async get(id) {
+    let found = await companyRepository.get(id);
```
Comment (lgnstn) [inline app/services/company.js]: This should be named `company`.

File: app/services/company.js [hunk 2]
```diff
@@ -3,30 +3,93 @@ const path = require('path');
...
+    if (!found) {
+      error.throwNotFoundError('Company not found.');
+    }
+
+    const { fieldOfActivityId, ...company } = found;
+    const fieldOfActivity =
```
Comment (lgnstn) [inline app/services/company.js]: Please indent it this way:
```
const fieldOfActivity = await fieldOfActivityRepository
    .get(fieldOfActivityId);
```

File: app/services/company.js [hunk 3]
```diff
@@ -3,30 +3,93 @@ const path = require('path');
...
+      name: company.name,
+    });
     if (companyExists) {
       error.throwValidationError('Company already exists.');
     }
-
-    if (company.fieldOfActivityId !== undefined) {
-      const fieldExists = await fieldOfActivityService.get(company.fieldOfActivityId);
+    const { business, address, ...newCompany } = company;
+    if (business !== undefined) {
```
Comment (lgnstn) [inline app/services/company.js]: This validation logic is taking too much space in this service. Please create another class that handles the validation.

File: app/services/company.js [hunk 4]
```diff
@@ -3,30 +3,93 @@ const path = require('path');
...
+    const createdCompany = await companyRepository.create(newCompany);
+
+    if (address !== undefined) {
+      const addressFields = address.split(', ');
+      const [ streetName, streetNumber ] = addressFields[0].split(' ');
+      const apartmentNumber = addressFields[1].split(' ')[1];
+      const city = addressFields[2];
+      const country = addressFields[3];
+
+      await addressRepository.create({
```
Comment (lgnstn) [inline app/services/company.js]: Please create a service for the Address if there isn't one already and call it here.

File: config.js [hunk 5]
```diff
@@ -1,4 +1,5 @@
 const tmpFolder = `${__dirname}/tmp`;
 const companyFilesFolder = `${__dirname}/files/companies`;
+const ORIGIN = 'http://localhost:3001/';
 
-module.exports = { tmpFolder, companyFilesFolder };
+module.exports = { tmpFolder, companyFilesFolder, ORIGIN };
```
Comment (lgnstn) [inline config.js]: Please use CAPITAL letters for all the constant exported here.

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Comment by laurandarius** [comment]
@lgnstn updated.

File: app/services/company.js [hunk 6]
```diff
@@ -1,43 +1,91 @@
...
-      if (!fieldOfActivity) {
-        error.throwNotFoundError('Field of activity not found.');
-      }
-      newCompany.fieldOfActivityId = fieldOfActivity.id;
-    } else {
-      error.throwValidationError('Invalid company format.');
+    const fieldOfActivity = await fieldOfActivityRepository.getByPropsNonParanoid({
+      name: business,
+    });
+    if (!fieldOfActivity) {
```
Comment (lgnstn) [inline app/services/company.js]: You can move this check in the validator as well.

File: app/services/company.js [hunk 7]
```diff
@@ -1,43 +1,91 @@
...
+      error.throwNotFoundError('Field of activity not found.');
     }
+    newCompany.fieldOfActivityId = fieldOfActivity.id;
 
     const createdCompany = await this.repository.create(newCompany);
 
+    if (address !== undefined) {
```
Comment (lgnstn) [inline app/services/company.js]: This whole block should be inside the address service.

File: app/services/company.js [hunk 8]
```diff
@@ -61,35 +109,43 @@ class CompanyService extends CrudService {
...
+        name: business,
+      });
       if (!fieldOfActivity) {
         error.throwNotFoundError('Field of activity not found.');
       }
       newData.fieldOfActivityId = fieldOfActivity.id;
     }
 
-    await this.get(id);
+    const companyExists = await this.repository.get(id);
```
Comment (lgnstn) [inline app/services/company.js]: 1) `companyExists` makes me think this is a boolean value, but in fact it is a company. Please rename it.
2) this check should be in the validator.

File: app/services/company.js [hunk 9]
```diff
@@ -61,35 +109,43 @@ class CompanyService extends CrudService {
...
+      error.throwNotFoundError('Company not found.');
+    }
     await this.repository.update(id, newData);
 
     return await this.get(id);
   }
 
   async delete(id) {
-    await this.get(id);
+    const companyExists = await this.repository.get(id);
```
Comment (lgnstn) [inline app/services/company.js]: Same.

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Comment by laurandarius** [comment]
@lgnstn updated.

File: app/services/address.js [hunk 10]
```diff
@@ -1,20 +1,39 @@
...
+  }
+
   async create(address) {
     if (address.id || address.companyId === undefined) {
       error.throwValidationError('Invalid address format.');
     }
 
-    const addressExists = await addressRepository.getByPropsNonParanoid(address);
-    if (addressExists) {
+    const existingAddress = await addressRepository.getByPropsNonParanoid(address);
```
Comment (lgnstn) [inline app/services/address.js]: Please create a validator for the addresses too.

File: app/services/company.js [hunk 11]
```diff
@@ -1,56 +1,90 @@
...
+    });
+    if (companyAddresses && companyAddresses.length) {
+      const {
+        streetName,
+        streetNumber,
+        apartmentNumber,
+        city,
+        country,
+      } = companyAddresses[0];
+      companyToReturn.address = `${streetName} ${streetNumber}, ap. ${apartmentNumber}, ${city}, ${country}`;
```
Comment (lgnstn) [inline app/services/company.js]: It is better to return the address object (model) and let the front end decide how it wants to display the address.

File: app/validators/company.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    if (data.name !== undefined) {
+      error.throwValidationError('You can not change the company name.');
+    }
+
+    const existingCompany = await companyRepository.get(id);
+    if (!existingCompany) {
+      error.throwNotFoundError('Company not found.');
+    }
+
+    const {
```
Comment (lgnstn) [inline app/validators/company.js]: The same code can be found in the method above. This is not ok, the code is not DRY (Don't Repeat Yourself). Please move it in a method and call it in both places.

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Comment by laurandarius** [comment]
@lgnstn updated.

File: app/services/address.js [hunk 13]
```diff
@@ -1,32 +1,35 @@
...
+    });
+  }
 
-    await companyService.get(address.companyId);
-    const createdAddress = await addressRepository.create(address);
+  async create(address) {
+    await addressValidator.validateCreate(address);
+    const createdAddress = await this.repository.create(address);
 
     return await this.get(createdAddress.id);
```
Comment (lgnstn) [inline app/services/address.js]: You not return here `createdAddress`?

File: app/services/address.js [hunk 14]
```diff
@@ -35,4 +38,5 @@ class AddressService extends CrudService {
 module.exports = new AddressService({
   repositoryName: 'address',
   modelName: 'Address',
+  validatorName: 'address',
```
Comment (lgnstn) [inline app/services/address.js]: Why do you need this? It doesn't look like you use it.

File: app/services/company.js [hunk 15]
```diff
@@ -1,56 +1,83 @@
...
-      const fieldOfActivity = await fieldOfActivityRepository.getByPropsNonParanoid({ name: business });
-      if (!fieldOfActivity) {
-        error.throwNotFoundError('Field of activity not found.');
-      }
-      newCompany.fieldOfActivityId = fieldOfActivity.id;
-    } else {
-      error.throwValidationError('Invalid company format.');
-    }
+    const fieldOfActivity = await fieldOfActivityRepository.getByPropsNonParanoid({
+      name: business,
```
Comment (lgnstn) [inline app/services/company.js:56]: You should have sent the field of activity id from the frontend. Keep this in mind in the future, don't change anything here now.

**Review by lgnstn — CHANGES_REQUESTED** [review]

**Comment by laurandarius** [comment]
@lgnstn updated.

**Review by lgnstn — APPROVED** [review]

**Pull request merged**